src/transformers/modeling_utils.py

```diff
@@ -1073,24 +1079,34 @@ def _generate_beam_search(
             if past:
                 past = self._reorder_cache(past, beam_idx)
 
-            # update current length
-            cur_len = cur_len + 1
-
             # stop when we are done with each sentence
             if all(done):
                 break
 
+            # update current length
+            cur_len = cur_len + 1
+
+        # finalize all open beam hypotheses and end to generated hypotheses
         for batch_idx in range(batch_size):
-            # Add all open beam hypothesis to generated_hyps
-            if not done[batch_idx]:
-                for idx, score in zip(next_tokens[batch_idx], next_scores[batch_idx]):
+            if done[batch_idx]:
+                continue
 
-                    # get beam and word IDs
-                    beam_id = idx // vocab_size
-                    token_id = idx % vocab_size
-                    generated_hyps[batch_idx].add(
-                        input_ids[batch_idx * num_beams + beam_id, :cur_len].clone(), score.item()
-                    )
+            # test that beam scores match previously calculated scores if not eos and batch_idx not done
```

Comment:
This refactoring was taken from bart's `generate()` function (@sshleifer) I think it's much cleaner by showing clearly that at the end the current best 3 (first 3) open hypotheses are added to the generated hypotheses. Also we take the final scores from the variable `beam_scores` here instead of `next_scores` (as before) which is the "correct" variable to take the scores from since it's the most current updated accumulated score. 
Also an assert statement verifying the `beam_scores` are the correctly calculated is added making sure that the logic will not be broken in further changes. @thomwolf @LysandreJik